doc/designs/quic-design/quic-api.md

```diff
@@ -0,0 +1,810 @@
+protocol concept
+
+Cons
+
+* Full QUIC will look very different to SSL
+* Writing a QUIC application is completely different to writing a TLS
+application - no easy path from one to the other
+
+
+# Solution Outline: Use the existing SSL API
+
+This section focuses on the "Just use the existing SSL API" candidate solution
+above and further elaborates a design for how that approach might work.
+
+The proposed additions to the ssl.h and bio.h header files for this approach
+are given in [Appendix A](#appendix-a).
+
+A key problem to solve with this design approach is how will QUIC events be
+delivered to an application? In a (D)TLS application there is a one-to-one
+correspondence between a socket and an `SSL` object. So an application can test
+for a socket becoming readable/writeable and then call `SSL_read()`/`SSL_write()`
+or some other `SSL` function in response. For servers it can also "accept" new
+connections on a socket to learn about new connection events.
+
+With QUIC there will be a single socket that is shared between multiple `SSL`
+objects (or possibly a set of sockets that are shared). A socket becoming
+readable/writeable can no longer be associated with a single `SSL` object. New
+connections are just new packets arriving on the socket(s) just like any other.
+There will need to be a mechanism to deliver events to an application that are
+associated with an individual `SSL` object in order for the standard `SSL_read()`/
+`SSL_write(`) functions, or some other IO function to be called.
+
+To solve this problem this design introduces a new `SSL_EVENT_CTX` object. An
+individual `SSL` object represents an individual stream. A set of `SSL` objects
+together represent all of the streams available for a given connection. The
+first `SSL` object created for a connection represents the connection itself as
+well as the first stream if there is one. All the `SSL` objects from all
+connections associated with a shared set of sockets are all grouped together by
+a single `SSL_EVENT_CTX`.
+
+While the principal motivation for introducing `SSL_EVENT_CTX` is to support QUIC,
+it may also be useful for supporting (D)TLS applications. Using an `SSL_EVENT_CTX`
+will be required for QUIC applications, but using one should (if desired by the
+application developer) also be possible for (D)TLS applications. It should also
+be possible to write (D)TLS applications in the traditional way.
+
+OpenSSL (D)TLS applications abstract away the details of the underlying
+transport protocol using `BIO`s. Application developers can either use one of the
+set of standard `BIO`s, or alternatively develop their own custom `BIO`s in order
+to have more control over the interaction with the underlying transport
+protocol. This approach needs to be possible with QUIC too. However for QUIC it
+will no longer be the case that each `SSL` object needs different `BIO` objects.
+Instead there will be common `BIO`s that are shared between multiple `SSL` objects.
```

Comment:
Are these stateless BIO's?

Reply:
Depends what you mean by stateless. The BIO structure has state in the them, and some functions might modify that state. However, the new functions `BIO_sendmmsg` and `BIO_recvmmsg` are intended to be callable statelessly, i.e. they do not modify the state of the underlying BIO and can be called simultaneously from multiple threads.